sound/soc/sof/ipc4-pcm.c

```diff
@@ -403,10 +404,56 @@ static int sof_ipc4_pcm_close(struct snd_soc_component *component,
 	return 0;
 }
 
+static snd_pcm_sframes_t sof_ipc4_pcm_delay(struct snd_soc_component *component,
+					    struct snd_pcm_substream *substream)
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct sof_ipc4_timestamp_info *time_info;
+	struct sof_ipc4_llp_reading_slot llp;
+	struct snd_sof_pcm_stream *stream;
+	struct snd_sof_pcm *spcm;
+	u64 link_pos;
+
+	spcm = snd_sof_find_spcm_dai(component, rtd);
+	if (!spcm)
+		return 0;
+
+	stream = &spcm->stream[substream->stream];
+	time_info = stream->time_info;
+
+	if (time_info->stream_start_offset == SOF_IPC4_INVALID_STREAM_POSITION) {
+		struct sof_ipc4_pipeline_registers ppl_reg;
+
+		sof_mailbox_read(sdev, sdev->fw_info_box.offset + time_info->ppl_reg_offset,
+				 &ppl_reg, sizeof(ppl_reg));
+
+		if (ppl_reg.stream_start_offset == SOF_IPC4_INVALID_STREAM_POSITION) {
+			dev_err(sdev->dev, "failed to get stream start offset\n");
+			return 0;
+		}
+
+		time_info->stream_start_offset = ppl_reg.stream_start_offset;
+	}
+
+	if (time_info->llp_offset)
+		sof_mailbox_read(sdev, time_info->llp_offset, &llp, sizeof(llp));
+
+	sof_ops(sdev)->pcm_pointer(sdev, substream);
+
+	if (!time_info->llp_offset)
+		link_pos = stream->posn.dai_posn;
+	else
+		link_pos = PACK_TWO_U32_TO_U64(llp.reading.llp_l, llp.reading.llp_u);
+
+	return stream->posn.host_posn - (link_pos - time_info->stream_start_offset);
```

Comment:
Not able to follow the arithmetic since the pcm_pointer doesn't rely on DPIB.

I will have to ask for a small paragraph describing how you compute all this, this is beyond confusing for reviewers.

Reply:
DPIB is register of 32bit counter, and the host_posn is at 64bit count. Kai commented that 32bit counter is not a good condition so I switched to 64bit counter

Reply:
@plbossart what do think about 32bit or 64 bit ?

Reply:
we absolutely have to use 64 bits for any time management. at 48kHz sample counters wrap around in 24.855134815 hours (2^32/(48000*3600)).

Reply:
DPIB is just a position pointer so 32 bits is just fine. We've not going to have a buffer needing more than 32-bits. But that position needs to be converted to a time and that's where 64bits comes in.

I must admit I am a bit surprised to see you call the ->pointer callback directly from the .delay callback. I am not sure if this is correct but I don't have time today to look at the details in pcm_native.c and pcm_lib.c.

Reply:
I think 32bit is ok as well. Just noting that 32bit counter will wrap around the code needs to handle the wrap-around condition.